File: src/screens/ExercisesScreen.js
import { useMemo, useState } from 'react';
import { View, Text, StyleSheet, SectionList, TouchableOpacity, TextInput } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import { getLibrary, getCategories } from '../data/exercises';
import { categoryIcon } from '../data/icons';
import { useTheme } from '../ThemeContext';
import { useDiscipline } from '../DisciplineContext';
import DisciplineSwitch from '../components/DisciplineSwitch';
import { spacing, radius } from '../theme';

export default function ExercisesScreen({ navigation }) {
  const { theme } = useTheme();
  const { discipline } = useDiscipline();
  const styles = useMemo(() => makeStyles(theme.colors), [theme]);
  const [query, setQuery] = useState('');

  const library = getLibrary(discipline);
  const categories = getCategories(discipline);

  const sections = useMemo(() => {
    const q = query.trim().toLowerCase();
    return categories
      .map((cat) => ({
        title: cat,
        data: library.filter(
          (e) => e.category === cat && (!q || e.name.toLowerCase().includes(q))
        ),
      }))
      .filter((s) => s.data.length > 0);
  }, [library, categories, query]);

  const total = library.length;

  return (
    <View style={styles.container}>
      <DisciplineSwitch />
      <View style={styles.searchWrap}>
        <TextInput
          style={styles.search}
          placeholder={`Cerca tra ${total} esercizi...`}
          placeholderTextColor={theme.colors.textMuted}
          value={query}
          onChangeText={setQuery}
        />
      </View>
      <SectionList
        sections={sections}
        keyExtractor={(item) => item.id}
        contentContainerStyle={styles.content}
        stickySectionHeadersEnabled={false}
        renderSectionHeader={({ section }) => (
          <View style={styles.headerRow}>
            <MaterialCommunityIcons
              name={categoryIcon(section.title)}
              size={20}
              color={theme.colors.primary}
            />
            <Text style={styles.header}>
              {section.title} <Text style={styles.count}>({section.data.length})</Text>
            </Text>
          </View>
        )}
        renderItem={({ item }) => (
          <TouchableOpacity
            style={styles.row}
            onPress={() => navigation.navigate('ExerciseDetail', { exerciseId: item.id })}
          >
            <Text style={styles.name}>{item.name}</Text>
            <View style={styles.metaRow}>
              <View style={styles.meta}>
                <Text style={styles.level}>{item.level}</Text>
                <Text style={styles.unit}>
                  {item.unit === 'weight' ? 'kg×reps' : item.unit}
                </Text>
              </View>
              {item.unit === 'sec' && (
                <TouchableOpacity
                  style={styles.timerBtn}
                  onPress={() => navigation.navigate('Timer', { name: item.name })}
                >
                  <Text style={styles.timerBtnText}>⏱️</Text>
                </TouchableOpacity>
              )}
              <Text style={styles.chevron}>›</Text>
            </View>
          </TouchableOpacity>
        )}
        ListEmptyComponent={
          <Text style={styles.empty}>Nessun esercizio trovato.</Text>
        }
      />
    </View>
  );
}

const makeStyles = (c) =>
  StyleSheet.create({
    container: { flex: 1, backgroundColor: c.bg },
    searchWrap: { padding: spacing.md, paddingBottom: 0 },
    search: {
      backgroundColor: c.card,
      color: c.text,
      borderRadius: radius.md,
      paddingHorizontal: spacing.md,
      paddingVertical: spacing.sm,
      borderWidth: 1,
      borderColor: c.border,
    },
    content: { padding: spacing.md },
    headerRow: {
      flexDirection: 'row',
      alignItems: 'center',
      gap: spacing.sm,
      marginTop: spacing.md,
      marginBottom: spacing.sm,
    },
    header: {
      color: c.primary,
      fontSize: 18,
      fontWeight: '700',
    },
    count: { color: c.textMuted, fontSize: 13, fontWeight: '400' },
    row: {
      flexDirection: 'row',
      justifyContent: 'space-between',
      alignItems: 'center',
      backgroundColor: c.card,
      borderRadius: radius.md,
      padding: spacing.md,
      marginBottom: spacing.sm,
      borderWidth: 1,
      borderColor: c.border,
    },
    name: { color: c.text, fontWeight: '600', flex: 1 },
    metaRow: { flexDirection: 'row', alignItems: 'center', gap: spacing.sm },
    meta: { alignItems: 'flex-end' },
    level: { color: c.textMuted, fontSize: 12 },
    unit: { color: c.textMuted, fontSize: 11 },
    timerBtn: {
      backgroundColor: c.cardAlt,
      borderRadius: radius.sm,
      paddingHorizontal: spacing.sm,
      paddingVertical: spacing.xs,
    },
    timerBtnText: { fontSize: 18 },
    chevron: { color: c.textMuted, fontSize: 22, fontWeight: '300' },
    empty: { color: c.textMuted, textAlign: 'center', marginTop: spacing.xl },
  });

File: lib/screens/exercises_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../state/theme_provider.dart';
import '../state/discipline_provider.dart';
import '../state/locale_provider.dart';
import '../data/exercises.dart';
import '../data/category_icons.dart';
import '../models/exercise.dart';
import '../theme/app_theme.dart';
import '../widgets/discipline_switch.dart';
import '../widgets/design_system.dart';
import '../modules/skills/skills_screen.dart';
import 'exercise_detail_screen.dart';
import 'timer_screen.dart';

class ExercisesScreen extends StatefulWidget {
  const ExercisesScreen({super.key});

  @override
  State<ExercisesScreen> createState() => _ExercisesScreenState();
}

class _ExercisesScreenState extends State<ExercisesScreen> {
  String _query = '';

  @override
  Widget build(BuildContext context) {
    final c = context.watch<ThemeProvider>().colors;
    final t = context.watch<LocaleProvider>().t;
    final discipline = context.watch<DisciplineProvider>().discipline;
    final categories = getCategories(discipline);
    final library = getLibrary(discipline);
    final filtered = _query.isEmpty
        ? library
        : library
              .where((e) => e.name.toLowerCase().contains(_query.toLowerCase()))
              .toList();

    return Scaffold(
      appBar: AppBar(title: Text(t('library'))),
      body: Column(
        children: [
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: Spacing.md),
            child: const DisciplineSwitch(),
          ),
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: Spacing.md),
            child: TextField(
              onChanged: (v) => setState(() => _query = v),
              decoration: InputDecoration(
                hintText: t.p('search_exercises', {'n': '${library.length}'}),
                prefixIcon: const Icon(Icons.search),
                isDense: true,
                border: OutlineInputBorder(
                  borderRadius: BorderRadius.circular(Radii.md),
                ),
              ),
            ),
          ),
          const SizedBox(height: Spacing.sm),
          Expanded(
            child: ListView(
              padding: const EdgeInsets.symmetric(horizontal: Spacing.md),
              children: [
                if (discipline == 'calisthenics') ...[
                  _SkillTreeBanner(
                    onTap: () => Navigator.push(
                      context,
                      MaterialPageRoute(builder: (_) => const SkillsScreen()),
                    ),
                  ),
                  const SizedBox(height: Spacing.sm),
                ],
                for (final cat in categories)
                  ..._buildCategory(context, cat, filtered, c),
                const SizedBox(height: 40),
              ],
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> _buildCategory(
    BuildContext context,
    String cat,
    List<Exercise> list,
    AppColors c,
  ) {
    final items = list.where((e) => e.category == cat).toList();
    if (items.isEmpty) return [];
    return [
      Padding(
        padding: const EdgeInsets.symmetric(vertical: Spacing.sm),
        child: Row(
          children: [
            Icon(categoryIcon(cat), color: c.primary, size: 18),
            const SizedBox(width: 8),
            Text(
              '$cat (${items.length})',
              style: TextStyle(color: c.textMuted, fontWeight: FontWeight.w700),
            ),
          ],
        ),
      ),
      ResponsiveWrap(
        minTileWidth: 340,
        maxColumns: 2,
        children: items
            .map(
              (ex) => Card(
                color: c.card,
                margin: EdgeInsets.zero,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(Radii.md),
                  side: BorderSide(color: c.border),
                ),
                child: ListTile(
                  leading: CircleAvatar(
                    backgroundColor: c.primary.withValues(alpha: 0.14),
                    child: Icon(
                      categoryIcon(ex.category),
                      color: c.primary,
                      size: 20,
                    ),
                  ),
                  title: Text(ex.name),
                  subtitle: Text(
                    _meta(ex),
                    style: TextStyle(color: c.textMuted),
                  ),
                  trailing: ex.unit == 'sec'
                      ? IconButton(
                          icon: Icon(Icons.timer_outlined, color: c.primary),
                          onPressed: () => Navigator.push(
                            context,
                            MaterialPageRoute(
                              builder: (_) =>
                                  TimerScreen(exerciseName: ex.name),
                            ),
                          ),
                        )
                      : Icon(Icons.chevron_right, color: c.textMuted),
                  onTap: () => Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (_) => ExerciseDetailScreen(exerciseId: ex.id),
                    ),
                  ),
                ),
              ),
            )
            .toList(),
      ),
    ];
  }

  String _meta(Exercise ex) {
    final unit = ex.unit == 'weight'
        ? 'kg×reps'
        : ex.unit == 'sec'
        ? 'secondi'
        : 'ripetizioni';
    return '${ex.level} · $unit';
  }
}

/// Banner d'ingresso all'Albero delle Skill.
class _SkillTreeBanner extends StatelessWidget {
  final VoidCallback onTap;
  const _SkillTreeBanner({required this.onTap});

  @override
  Widget build(BuildContext context) {
    final c = context.watch<ThemeProvider>().colors;
    final t = context.watch<LocaleProvider>().t;
    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding: const EdgeInsets.all(Spacing.md),
        decoration: BoxDecoration(
          gradient: LinearGradient(
            colors: [
              Color.alphaBlend(c.primary.withValues(alpha: 0.2), c.card),
              c.card,
            ],
          ),
          borderRadius: BorderRadius.circular(Radii.md),
          border: Border.all(color: c.primary.withValues(alpha: 0.4)),
        ),
        child: Row(
          children: [
            Container(
              width: 42,
              height: 42,
              decoration: BoxDecoration(
                color: c.primary.withValues(alpha: 0.18),
                borderRadius: BorderRadius.circular(Radii.md),
              ),
              child: Icon(Icons.account_tree_rounded, color: c.primary),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Text(
                t('skills_open'),
                style: const TextStyle(
                  fontSize: 15,
                  fontWeight: FontWeight.w800,
                ),
              ),
            ),
            Icon(Icons.chevron_right_rounded, color: c.textMuted),
          ],
        ),
      ),
    );
  }
}
